File: src/VisualBasic/CodeCracker/Refactoring/ChangeAnyToAllAnalyzer.vb
Imports System.Collections.Immutable
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Diagnostics
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Refactoring
    <DiagnosticAnalyzer(LanguageNames.VisualBasic)>
    Public Class ChangeAnyToAllAnalyzer
        Inherits DiagnosticAnalyzer

        Friend Const MessageAny = "Change Any to All"
        Friend Const MessageAll = "Change All to Any"
        Friend Const TitleAny = MessageAny
        Friend Const TitleAll = MessageAll
        Friend Const Category = SupportedCategories.Refactoring

        Friend Shared RuleAny As New DiagnosticDescriptor(
            DiagnosticId.ChangeAnyToAll.ToDiagnosticId(),
            TitleAny,
            MessageAny,
            Category,
            DiagnosticSeverity.Hidden,
            isEnabledByDefault:=True,
            helpLinkUri:=HelpLink.ForDiagnostic(DiagnosticId.ChangeAnyToAll))
        Friend Shared RuleAll As New DiagnosticDescriptor(
            DiagnosticId.ChangeAllToAny.ToDiagnosticId(),
            TitleAll,
            MessageAll,
            Category,
            DiagnosticSeverity.Hidden,
            isEnabledByDefault:=True,
            helpLinkUri:=HelpLink.ForDiagnostic(DiagnosticId.ChangeAllToAny))

        Public Overrides ReadOnly Property SupportedDiagnostics As ImmutableArray(Of DiagnosticDescriptor)
            Get
                Return ImmutableArray.Create(RuleAll, RuleAny)
            End Get
        End Property

        Public Overrides Sub Initialize(context As AnalysisContext)
            context.RegisterSyntaxNodeAction(AddressOf AnalyzeInvocation, SyntaxKind.InvocationExpression)
        End Sub

        Public Shared ReadOnly allName As IdentifierNameSyntax = SyntaxFactory.IdentifierName(NameOf(System.Linq.Enumerable.All))
        Public Shared ReadOnly anyName As IdentifierNameSyntax = SyntaxFactory.IdentifierName(NameOf(System.Linq.Enumerable.Any))

        Private Sub AnalyzeInvocation(context As SyntaxNodeAnalysisContext)
            If context.IsGenerated() Then Exit Sub
            Dim invocation = DirectCast(context.Node, InvocationExpressionSyntax)
            If invocation.Parent.IsKind(SyntaxKind.ExpressionStatement) Then Exit Sub
            Dim diagnosticToRaise = GetCorrespondingDiagnostic(context.SemanticModel, invocation)
            If diagnosticToRaise Is Nothing Then Exit Sub
            Dim diag = Diagnostic.Create(diagnosticToRaise, DirectCast(invocation.Expression, MemberAccessExpressionSyntax).Name.GetLocation())
            context.ReportDiagnostic(diag)
        End Sub

        Private Shared Function GetCorrespondingDiagnostic(model As SemanticModel, invocation As InvocationExpressionSyntax) As DiagnosticDescriptor
            Dim methodName = TryCast(invocation?.Expression, MemberAccessExpressionSyntax)?.Name?.ToString()
            Dim nameToCheck = If(methodName = NameOf(System.Linq.Enumerable.Any), allName,
                If(methodName = NameOf(System.Linq.Enumerable.All), anyName, Nothing))
            If nameToCheck Is Nothing Then Return Nothing
            Dim invocationSymbol = TryCast(model.GetSymbolInfo(invocation).Symbol, IMethodSymbol)
            If invocationSymbol?.Parameters.Length <> 1 Then Return Nothing
            If Not IsLambdaWithEmptyOrSimpleBody(invocation) Then Return Nothing
            Dim otherInvocation = invocation.WithExpression(DirectCast(invocation.Expression, MemberAccessExpressionSyntax).WithName(nameToCheck))
            Dim otherInvocationSymbol = model.GetSpeculativeSymbolInfo(invocation.SpanStart, otherInvocation, SpeculativeBindingOption.BindAsExpression)
            If otherInvocationSymbol.Symbol Is Nothing Then Return Nothing
            Return If(methodName = NameOf(System.Linq.Enumerable.Any), RuleAny, RuleAll)
        End Function

        Private Shared Function IsLambdaWithEmptyOrSimpleBody(invocation As InvocationExpressionSyntax) As Boolean
            Dim arg = invocation.ArgumentList?.Arguments.FirstOrDefault()
            If arg Is Nothing Then Return True ' Empty body
            Return arg.GetExpression().Kind = SyntaxKind.SingleLineFunctionLambdaExpression
        End Function
    End Class
End Namespace

File: src/CSharp/CodeCracker/Refactoring/ChangeAnyToAllAnalyzer.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;
using System.Collections.Immutable;
using System.Linq;

namespace CodeCracker.CSharp.Refactoring
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class ChangeAnyToAllAnalyzer : DiagnosticAnalyzer
    {
        private const string speculativeAnnotationDescription = "ChangeAnyToAllAnalyzer_speculativeAnnotation";
        private static readonly SyntaxAnnotation speculativeAnnotation = new SyntaxAnnotation(speculativeAnnotationDescription);
        internal const string MessageAny = "Change Any to All";
        internal const string MessageAll = "Change All to Any";
        internal const string TitleAny = MessageAny;
        internal const string TitleAll = MessageAll;
        internal const string Category = SupportedCategories.Refactoring;

        internal static readonly DiagnosticDescriptor RuleAny = new DiagnosticDescriptor(
            DiagnosticId.ChangeAnyToAll.ToDiagnosticId(),
            TitleAny,
            MessageAny,
            Category,
            SeverityConfigurations.CurrentCS[DiagnosticId.ChangeAnyToAll],
            isEnabledByDefault: true,
            helpLinkUri: HelpLink.ForDiagnostic(DiagnosticId.ChangeAnyToAll));
        internal static readonly DiagnosticDescriptor RuleAll = new DiagnosticDescriptor(
            DiagnosticId.ChangeAllToAny.ToDiagnosticId(),
            TitleAll,
            MessageAll,
            Category,
            SeverityConfigurations.CurrentCS[DiagnosticId.ChangeAllToAny],
            isEnabledByDefault: true,
            helpLinkUri: HelpLink.ForDiagnostic(DiagnosticId.ChangeAllToAny));

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(RuleAny, RuleAll);

        public override void Initialize(AnalysisContext context) => context.RegisterSyntaxNodeAction(AnalyzeInvocation, SyntaxKind.InvocationExpression);

        public static readonly SimpleNameSyntax allName = (SimpleNameSyntax)SyntaxFactory.ParseName("All");
        public static readonly SimpleNameSyntax anyName = (SimpleNameSyntax)SyntaxFactory.ParseName("Any");

        private static void AnalyzeInvocation(SyntaxNodeAnalysisContext context)
        {
            if (context.IsGenerated()) return;
            var invocation = (InvocationExpressionSyntax)context.Node;
            if (invocation.Parent?.IsKind(SyntaxKind.ExpressionStatement) ?? true) return;
            var diagnosticToRaise = GetCorrespondingDiagnostic(context.SemanticModel, invocation);
            if (diagnosticToRaise == null) return;
            var diagnostic = Diagnostic.Create(diagnosticToRaise, GetName(invocation).GetLocation());
            context.ReportDiagnostic(diagnostic);
        }

        private static DiagnosticDescriptor GetCorrespondingDiagnostic(SemanticModel semanticModel, InvocationExpressionSyntax invocation)
        {
            var methodName = GetName(invocation);
            var methodNameText = methodName?.ToString();
            var nameToCheck = methodNameText == "Any" ? allName : methodNameText == "All" ? anyName : null;
            if (nameToCheck == null) return null;
            var methodSymbol = semanticModel.GetSymbolInfo(invocation.Expression).Symbol as IMethodSymbol;
            if (methodSymbol?.Parameters.Length != 1) return null;
            if (!IsLambdaWithoutBody(invocation)) return null;
            if (!OtherMethodExists(invocation, nameToCheck, semanticModel)) return null;
            return methodNameText == "Any" ? RuleAny : RuleAll;
        }

        public static SimpleNameSyntax GetName(InvocationExpressionSyntax invocation)
        {
            SimpleNameSyntax methodName = null;
            if (invocation.Expression.IsKind(SyntaxKind.MemberBindingExpression))
                methodName = ((MemberBindingExpressionSyntax)invocation.Expression).Name;
            else if (invocation.Expression.IsKind(SyntaxKind.SimpleMemberAccessExpression))
                methodName = ((MemberAccessExpressionSyntax)invocation.Expression).Name;
            return methodName;
        }

        private static bool IsLambdaWithoutBody(InvocationExpressionSyntax invocation)
        {
            var arg = invocation.ArgumentList?.Arguments.First();
            var lambda = arg.Expression as LambdaExpressionSyntax;
            if (lambda == null) return false;
            return !(lambda.Body is BlockSyntax);
        }

        private static bool OtherMethodExists(InvocationExpressionSyntax invocation, SimpleNameSyntax nameToCheck, SemanticModel semanticModel)
        {
            var otherExpression = CreateExpressionWithNewName(invocation, nameToCheck);
            var statement = invocation.FirstAncestorOrSelfThatIsAStatement();
            SemanticModel speculativeModel;
            if (statement != null)
            {
                var otherStatement = statement.ReplaceNode(invocation.Expression, otherExpression);
                if (!semanticModel.TryGetSpeculativeSemanticModel(statement.SpanStart, otherStatement, out speculativeModel)) return false;
            }
            else
            {
                var arrow = (ArrowExpressionClauseSyntax)invocation.FirstAncestorOfKind(SyntaxKind.ArrowExpressionClause);
                if (arrow == null) return false;
                var otherArrow = arrow.ReplaceNode(invocation.Expression, otherExpression);
                if (!semanticModel.TryGetSpeculativeSemanticModel(arrow.SpanStart, otherArrow, out speculativeModel)) return false;
            }
            var symbol = speculativeModel.GetSymbolInfo(speculativeModel.SyntaxTree.GetRoot().GetAnnotatedNodes(speculativeAnnotationDescription).First()).Symbol;
            return symbol != null;
        }

        public static ExpressionSyntax CreateExpressionWithNewName(InvocationExpressionSyntax invocation, SimpleNameSyntax nameToCheck)
        {
            var otherExpression = invocation.Expression.IsKind(SyntaxKind.MemberBindingExpression)
                ? (ExpressionSyntax)((MemberBindingExpressionSyntax)invocation.Expression).WithName(nameToCheck).WithAdditionalAnnotations(speculativeAnnotation)
                //avoid this, already checked before: if (invocation.Expression.IsKind(SyntaxKind.SimpleMemberAccessExpression)):
                : ((MemberAccessExpressionSyntax)invocation.Expression).WithName(nameToCheck).WithAdditionalAnnotations(speculativeAnnotation);
            return otherExpression;
        }
    }
}